cirq-core/cirq/ops/gate_features.py

```diff
@@ -36,20 +36,36 @@ class SupportsOnEachGate(raw_types.Gate, metaclass=abc.ABCMeta):
 
     def on_each(self, *targets: Union[raw_types.Qid, Iterable[Any]]) -> List[raw_types.Operation]:
         """Returns a list of operations applying the gate to all targets.
-
         Args:
-            *targets: The qubits to apply this gate to.
-
+            *targets: The qubits to apply this gate to. For single-qubit gates
+            this can be provided as varargs or a combination of nested
+            iterables. For multi-qubit gates this must be provided as an
+            `Iterable[Sequence[Qid]]`, where each sequence has `num_qubits`
+            qubits.
         Returns:
             Operations applying this gate to the target qubits.
-
         Raises:
-            ValueError if targets are not instances of Qid or List[Qid].
-            ValueError if the gate operates on two or more Qids.
+            ValueError if targets are not instances of Qid or Iterable[Qid].
+            ValueError if the gate qubit number is incompatible.
         """
         if self._num_qubits_() > 1:
-            raise ValueError('This gate only supports on_each when it is a one qubit gate.')
-        operations = []  # type: List[raw_types.Operation]
+            ops: List[raw_types.Operation] = []
+            if len(targets) != 1 or not isinstance(targets[0], Iterable):
```

Comment:
I know there's been a lot of debate over the precise semantics: https://github.com/quantumlib/Cirq/issues/4034#issuecomment-867136721

My opinion would be to support option A and B. While flattening arbitrarily nested trees of qubits could monkey with the validation logic for multiqubit gates, both A and B are very explicit about which gates get which pair (or triple, quadruplet, ...) of qubits

Reply:
SGTM. I think I prefer that too.

Reply:
FWIW I'm sure there's a way to do it but somehow supporting A and B only is surprisingly complex. Two hours in trying multiple approaches and I've still got test failures. My guess is that's why the single-qubit case was implemented like it is.

Reply:
Okay, this is the simplest implementation I could figure out that worked for all cases. (Zip handling was rather challenging because if you looked at it to check the type of the things inside it, it didn't allow you to iterate again).